File: leetcode/lc-us/cpp/1519.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 1e5 + 5, M = N << 1;

bool st[N];
int h[N], e[M], ne[M], idx = 0;

class Solution {
    int hash[N][26];

    vector<int> dfs(string &s, int u) {
        vector<int> cnt(26);
        cnt[s[u] - 'a']++;
        for (int i = h[u]; ~i; i = ne[i]) {
            int j = e[i];
            if (!st[j]) {
                st[j] = true;
                auto t = dfs(s, j);
                for (int k = 0; k < 26; k++) {
                    cnt[k] += t[k];
                }
            }
        }
        for (int i = 0; i < 26; i++) {
            hash[u][i] = cnt[i];
        }
        return cnt;
    };
public:
    vector<int> countSubTrees(int n, vector<vector<int>> &edges, string labels) {
        idx = 0;
        memset(h, -1, sizeof h);
        memset(st, false, sizeof st);
        memset(hash, 0, sizeof hash);

        for (auto &edge: edges) {
            e[idx] = edge[1], ne[idx] = h[edge[0]], h[edge[0]] = idx++;
            e[idx] = edge[0], ne[idx] = h[edge[1]], h[edge[1]] = idx++;
        }
        vector<int> res(n);
        st[0] = true;
        dfs(labels, 0);
        for (int i = 0; i < n; i++) {
            res[i] = hash[i][labels[i] - 'a'];
        }
        return res;
    }
};

File: leetcode/lc-us/py/1519.py
from collections import defaultdict

from typing import List


class Solution:
    def countSubTrees(self, n: int, edges: List[List[int]], labels: str) -> List[int]:
        g = [[] for _ in range(n)]
        for e in edges:
            g[e[0]].append(e[1])
            g[e[1]].append(e[0])

        res = [1] * n

        def dfs(u: int, p: int) -> defaultdict[int]:
            cnt = defaultdict(int)
            cnt[labels[u]] += 1
            for x in g[u]:
                if x != p:
                    t = dfs(x, u)
                    for k, v in t.items():
                        cnt[k] += v
            res[u] = cnt[labels[u]]
            return cnt

        dfs(0, -1)
        return res
